components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/EmailDomainValidationHandler.java

```diff
@@ -0,0 +1,232 @@
+                return false;
+            } else {
+                organizationId = organizationManager.getPrimaryOrganizationId(organizationId);
+            }
+
+            return isEmailDomainDiscoveryEnabled(organizationId);
+        } catch (OrganizationConfigClientException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("No organization discovery configurations found for organization: " + CarbonContext
+                        .getThreadLocalCarbonContext().getTenantDomain());
+            }
+            return false;
+        } catch (OrganizationManagementException | OrganizationConfigException e) {
+            log.error("Error while retrieving organization discovery configuration.", e);
+            return false;
+        }
+    }
+
+    @Override
+    public int getPriority() {
+
+        int priority = super.getPriority();
+        if (priority == -1) {
+            priority = 15;
+        }
+        return priority;
+    }
+
+    @Override
+    public PostAuthnHandlerFlowStatus handle(HttpServletRequest request, HttpServletResponse response,
+                                             AuthenticationContext context) throws PostAuthenticationFailedException {
+
+        SequenceConfig sequenceConfig = context.getSequenceConfig();
+        for (Map.Entry<Integer, StepConfig> entry : sequenceConfig.getStepMap().entrySet()) {
+            StepConfig stepConfig = entry.getValue();
+            AuthenticatorConfig authenticatorConfig = stepConfig.getAuthenticatedAutenticator();
+            if (authenticatorConfig == null) {
+                continue;
+            }
+
+            ApplicationAuthenticator authenticator = authenticatorConfig.getApplicationAuthenticator();
+            if (authenticator instanceof FederatedApplicationAuthenticator) {
+                Map<String, String> localClaimValues;
+                if (stepConfig.isSubjectAttributeStep()) {
+                    localClaimValues =
+                            (Map<String, String>) context.getProperty(FrameworkConstants.UNFILTERED_LOCAL_CLAIM_VALUES);
+                } else {
+                    // Need to validate even if this is not the subject attribute step since
+                    // jit provisioning will happen in both scenarios.
+                    localClaimValues =
+                            FrameworkUtils.getLocalClaimValuesOfIDPInNonAttributeSelectionStep(context, stepConfig,
+                                    context.getExternalIdP());
+                }
+
+                String emailDomain = extractEmailDomain(localClaimValues.get(FrameworkConstants.EMAIL_ADDRESS_CLAIM));
+
+                if (emailDomain == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Email address not found or is not in the correct format." +
+                                " Email domain validation failed.");
```

Comment:
Is it possible to log a tenant domain or organization id here? If possible that will be helpful for the debugging purposes.